core/unit_test/tools/include/ToolTestingUtilities.hpp

```diff
@@ -0,0 +1,1175 @@
+
+KOKKOS_IMPL_TOOLS_TEST_CONFIG_OPTION(Kernels, profiling.kernels, 2);
+KOKKOS_IMPL_TOOLS_TEST_CONFIG_OPTION(Fences, profiling.fences, 2);
+KOKKOS_IMPL_TOOLS_TEST_CONFIG_OPTION(Allocs, profiling.allocs, 2);
+KOKKOS_IMPL_TOOLS_TEST_CONFIG_OPTION(Copies, profiling.copies, 2);
+KOKKOS_IMPL_TOOLS_TEST_CONFIG_OPTION(DualViewOps, profiling.dual_view_ops, 2);
+KOKKOS_IMPL_TOOLS_TEST_CONFIG_OPTION(Contexts, tuning.contexts, 2);
+KOKKOS_IMPL_TOOLS_TEST_CONFIG_OPTION(TypeDeclarations, tuning.type_declarations,
+                                     2);
+KOKKOS_IMPL_TOOLS_TEST_CONFIG_OPTION(RequestValues, tuning.request_values, 2);
+KOKKOS_IMPL_TOOLS_TEST_CONFIG_OPTION(Init, infrastructure.init, 2);
+KOKKOS_IMPL_TOOLS_TEST_CONFIG_OPTION(Finalize, infrastructure.finalize, 2);
+KOKKOS_IMPL_TOOLS_TEST_CONFIG_OPTION(ProgrammingInterface,
+                                     infrastructure.programming_interface, 2);
+KOKKOS_IMPL_TOOLS_TEST_CONFIG_OPTION(RequestSettings,
+                                     infrastructure.request_settings, 2);
+
+template <bool target_value>
+struct ToggleInfrastructure : public std::integral_constant<int, 1> {
+  void operator()(ToolValidatorConfiguration& config) {
+    config.infrastructure = {target_value, target_value, target_value,
+                             target_value};
+  }
+};
+
+using EnableInfrastructure  = ToggleInfrastructure<true>;
+using DisableInfrastructure = ToggleInfrastructure<false>;
+
+template <bool target_value>
+struct ToggleProfiling : public std::integral_constant<int, 1> {
+  void operator()(ToolValidatorConfiguration& config) {
+    config.profiling = {target_value, target_value, target_value, target_value,
+                        target_value};
+  }
+};
+
+using EnableProfiling  = ToggleProfiling<true>;
+using DisableProfiling = ToggleProfiling<false>;
+
+template <bool target_value>
+struct ToggleTuning : public std::integral_constant<int, 1> {
+  void operator()(ToolValidatorConfiguration& config) {
+    config.tuning = {target_value, target_value, target_value};
+  }
+};
+
+using EnableTuning  = ToggleTuning<true>;
+using DisableTuning = ToggleTuning<false>;
+
+template <bool target_value>
+struct ToggleAll : public std::integral_constant<int, 0> {
+  void operator()(ToolValidatorConfiguration& config) {
+    ToggleProfiling<target_value>{}(config);
+    ToggleTuning<target_value>{}(config);
+    ToggleInfrastructure<target_value>{}(config);
+  }
+};
+
+using EnableAll  = ToggleAll<true>;
+using DisableAll = ToggleAll<false>;
+}  // namespace Config
+
+/**
+ * Needs to stand outside of functions, this is the vector tool callbacks will
+ * push events into
+ */
+std::vector<EventBasePtr> found_events;
+/**
+ * Needs to stand outside of functions, this is the kID of the last encountered
+ * begin event
+ */
+static uint64_t last_kid;
+/** Subscribes to all of the requested callbacks */
+void set_tool_events_impl(ToolValidatorConfiguration& config) {
```

Comment:
We can't, unfortunately, though I'll leave a comment. It has to be usable in tools callbacks, which are function pointers. So it has to be a global :/. It does legitimately suck though

Reply:
Can't you pass (and use) a reference instead?

Reply:
I can't figure out a way. Here's my Godbolt on it

https://godbolt.org/z/xxbTdrjhs

But most phrasings of this will violate the rules of not capturing

Reply:
Something like https://godbolt.org/z/vq45eG7Wo?

Reply:
Problem is that it isn't a std::function that we use for callbacks, it's a raw function pointer, so &events is illegal

Reply:
https://godbolt.org/z/bdEo3bWKn (for reference)

Reply:
That's what I missed. I guess it's OK then.